url4fdf79ae2c_beckchr_juel_tgz-pJ8-de_odysseus_el_samples_xml_sax_XMELFilterJ8/src/javax/el/BeanELResolver.java

```diff
@@ -0,0 +1,666 @@
+import java.beans.PropertyDescriptor;
+import java.lang.reflect.Array;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Defines property resolution behavior on objects using the JavaBeans component architecture. This
+ * resolver handles base objects of any type, as long as the base is not null. It accepts any object
+ * as a property, and coerces it to a string. That string is then used to find a JavaBeans compliant
+ * property on the base object. The value is accessed using JavaBeans getters and setters. This
+ * resolver can be constructed in read-only mode, which means that isReadOnly will always return
+ * true and {@link #setValue(ELContext, Object, Object, Object)} will always throw
+ * PropertyNotWritableException. ELResolvers are combined together using {@link CompositeELResolver}
+ * s, to define rich semantics for evaluating an expression. See the javadocs for {@link ELResolver}
+ * for details. Because this resolver handles base objects of any type, it should be placed near the
+ * end of a composite resolver. Otherwise, it will claim to have resolved a property before any
+ * resolvers that come after it get a chance to test if they can do so as well.
+ * 
+ * @see CompositeELResolver
+ * @see ELResolver
+ */
+public class BeanELResolver extends ELResolver {
+	protected static final class BeanProperties {
+		private final Map<String, BeanProperty> map = new HashMap<String, BeanProperty>();
+
+		public BeanProperties(Class<?> baseClass) {
+			PropertyDescriptor[] descriptors;
+			try {
+				descriptors = Introspector.getBeanInfo(baseClass).getPropertyDescriptors();
+			} catch (IntrospectionException e) {
+				throw new ELException(e);
+			}
+			for (PropertyDescriptor descriptor : descriptors) {
+				map.put(descriptor.getName(), new BeanProperty(descriptor));
+			}
+		}
+
+		public BeanProperty getBeanProperty(String property) {
+			return map.get(property);
+		}
+	}
+
+	protected static final class BeanProperty {
+		private final PropertyDescriptor descriptor;
+		
+		private Method readMethod;
+		private Method writedMethod;
+
+		public BeanProperty(PropertyDescriptor descriptor) {
+			this.descriptor = descriptor;
+		}
+
+		public Class<?> getPropertyType() {
+			return descriptor.getPropertyType();
+		}
+
+		public Method getReadMethod() {
+			if (readMethod == null) {
+				readMethod = findAccessibleMethod(descriptor.getReadMethod());
+			}
+			return readMethod;
+		}
+
+		public Method getWriteMethod() {
+			if (writedMethod == null) {
+				writedMethod = findAccessibleMethod(descriptor.getWriteMethod());
+			}
+			return writedMethod;
+		}
+
+		public boolean isReadOnly() {
+			return getWriteMethod() == null;
+		}
+	}
+
+	private static Method findPublicAccessibleMethod(Method method) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 11. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 109-119, lines 120-131 into separate methods.